### DISCUSSION TITLE
How restore filter from UriQueryExpressionParser to string?

### DISCUSSION THREAD

**Discussion by P9avel** [OData/AspNetCoreOData#1105] [General] [answered]
Hi. I have code
`
var parser = new UriQueryExpressionParser(100);
var filter = parser.ParseFilter("city/name eq 'London'");
Console.WriteLine(filter);
`
How can i get source string "city/name eq 'London'" from filter variable?

**Comment by aboryczko** [comment, marked as answer]
What do you want to achieve? UriQueryExpressionParser will return a tree of tokens (Abstract Syntax Tree). You would need to visit the nodes in order to process it. The result of your ParseFilter would look something like this:

```mermaid
flowchart TD
    bo[BinaryOperatorToken Equals]-->ep;
    bo-->lt[LiteralToken London];
    ep[EndPathToken name]-->ip[InnerPathToken city];
```

You could implement the ISyntacticTreeVisitor interface to visit the different nodes and change it into a string. In the library I've seen that they use Binders to change QueryTokens to QueryNodes and there is a NodeToStringBuilder that builds the query string from the tree.

**Comment by P9avel** [comment]
Many thx for your answer
